### PR TITLE
4829: Display associated services on order form

### PR DESCRIPTION
**Pull request by SimonLLawrence**
Story [AB#4829](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/4829)
image

### CODE REVIEW

File: app/pages/task-list/helpers/generateTaskList.test.js [hunk 1]
```diff
@@ -198,6 +210,50 @@ describe('generateTaskList', () => {
...
+          {
+            id: 'task2item2',
+            status: 'complete',
+            count: 1,
+          },
+        ];
+        const taskList = generateTaskList({
+          orderId, taskListManifest, sectionsData: sectionDataWithCount,
+        });
+        expect(taskList[4].items[0].href).toBeFalsy();
```
Comment (alibombaye) [inline app/pages/task-list/helpers/generateTaskList.test.js:232]: 👍

File: app/pages/task-list/helpers/generateTaskList.test.js [hunk 2]
```diff
@@ -198,6 +210,50 @@ describe('generateTaskList', () => {
...
+            id: 'task2item2',
+            status: 'complete',
+            count: 0,
+          },
+        ];
+        const taskList = generateTaskList({
+          orderId, taskListManifest, sectionsData: sectionDataWithCount,
+        });
+        expect(taskList[4].items[0].href).toBeTruthy();
+      });
```
Comment (alibombaye) [inline app/pages/task-list/helpers/generateTaskList.test.js:255]: 👍

File: app/pages/task-list/taskListManifest.json [hunk 3]
```diff
@@ -81,6 +81,38 @@
...
+              "zeroCountDependencies": ["additional-services"],
+              "countDependencies": ["service-recipients","catalogue-solutions"]
+            },
+            { 
+              "statusDependencies": ["description", "ordering-party", "supplier", "commencement-date", "service-recipients","catalogue-solutions","additional-services"],
+              "countDependencies": ["service-recipients","catalogue-solutions","additional-services"]
+            },
+            { 
+              "statusDependencies": ["associated-services"]
+            }
```
Comment (alibombaye) [inline app/pages/task-list/taskListManifest.json:112]: don't think we need this one

Comment (SimonLLawrence) [reply app/pages/task-list/taskListManifest.json:112]: We were trying to cover Scenario 8

Scenario 8 - Select no Service Recipients and then go back and add service recipients
Given that a User indicates there are no Service Recipients in the order
and the Associated Service section is enabled
and the User adds an Associated Service
When the User goes back to the Service Recipient section
and adds a Service Recipient
Then the Associated Service section will remain enabled

Comment (SimonLLawrence) [reply app/pages/task-list/taskListManifest.json:112]: What way would you recommend we ensure that is covered?

File: app/pages/task-list/taskListManifest.json [hunk 4]
```diff
@@ -81,6 +81,38 @@
...
+            },
+            { 
+              "statusDependencies": ["description", "ordering-party", "supplier", "commencement-date", "service-recipients","catalogue-solutions","additional-services"],
+              "zeroCountDependencies": ["additional-services"],
+              "countDependencies": ["service-recipients","catalogue-solutions"]
+            },
+            { 
+              "statusDependencies": ["description", "ordering-party", "supplier", "commencement-date", "service-recipients","catalogue-solutions","additional-services"],
+              "countDependencies": ["service-recipients","catalogue-solutions","additional-services"]
+            },
```
Comment (alibombaye) [inline app/pages/task-list/taskListManifest.json:109]: ok little confused I thought there are 2 combos it is enabled. One when service recipient is 0
```
{
              "statusDependencies": ["description", "ordering-party", "supplier", "commencement-date", "service-recipients"],
              "zeroCountDependencies": ["service-recipients"]
            },
```
Which is what you have.

And then the other combo is when service recipient is > 0 then we need catalogue-solutions and additional services as complete for it to be enabled

```
{ 
              "statusDependencies": ["description", "ordering-party", "supplier", "commencement-date", "service-recipients","catalogue-solutions","additional-services"],
              "countDependencies": ["service-recipients",]
            },
```
^ this was the only other combo I was expecting for the section to be enabled

File: app/pages/task-list/taskListManifest.json (hunk 3 above)
Comment (alibombaye) [reply app/pages/task-list/taskListManifest.json:112]: 🤕

File: app/pages/task-list/taskListManifest.json (hunk 4 above)
Comment (SimonLLawrence) [reply app/pages/task-list/taskListManifest.json:109]: That is for AC 7 

Scenario 7: Additional Service now complete, >= 1 Service Recipient, >=1 Catalogue Solution, =>1 Additional Service

Its says the counts must be grater than zero ?

File: app/pages/task-list/taskListManifest.json (hunk 3 above)
Comment (alibombaye) [reply app/pages/task-list/taskListManifest.json:112]: ah man that scenario is a shambles..  but I think it'll be something like this

```
{ 
              "statusDependencies": ["description", "ordering-party", "supplier", "commencement-date", "service-recipients","associated-services"],
              "countDependencies": ["service-recipients", "associated-services"]
            },
```

Comment (SimonLLawrence) [reply app/pages/task-list/taskListManifest.json:112]: I need to investigate your recommended change, as it causes a tests to fail which passes with the other version. Will look into it.

Comment (SimonLLawrence) [reply app/pages/task-list/taskListManifest.json:112]: ok sorted that out

File: app/pages/task-list/taskListManifest.json (hunk 4 above)
Comment (SimonLLawrence) [reply app/pages/task-list/taskListManifest.json:109]: Changed as advised above

File: app/pages/task-list/existingorder/ui-tests/ui.test.js [hunk 5]
```diff
@@ -555,6 +555,210 @@ test('should only render the complete tag for task 6 item 1 when returned as com
...
+    { id: 'service-recipients', status: 'complete', count: 1 },
+  ]);
+  await pageSetup(t, true, mockOrderSummary);
+  await t.navigateTo(pageUrl);
+
+  const task7Item1 = Selector('li[data-test-id="task-6-item-0"]');
+
+  await t
+    .expect(task7Item1.find('a').exists).notOk();
+});
```
Comment (alibombaye) [inline app/pages/task-list/existingorder/ui-tests/ui.test.js:588]: 👍

File: app/pages/task-list/existingorder/ui-tests/ui.test.js [hunk 6]
```diff
@@ -555,6 +555,210 @@ test('should only render the complete tag for task 6 item 1 when returned as com
...
+  await pageSetup(t, true, mockOrderSummary);
+  await t.navigateTo(pageUrl);
+
+  const task7Item1 = Selector('li[data-test-id="task-6-item-0"]');
+
+  await t
+    .expect(task7Item1.find('a').exists).ok()
+    .click(task7Item1.find('a'))
+    .expect(getLocation()).eql(`http://localhost:1234${baseUrl}/organisation/order-id/associated-services`);
+});
```
Comment (alibombaye) [inline app/pages/task-list/existingorder/ui-tests/ui.test.js:606]: 👍

File: app/pages/task-list/existingorder/ui-tests/ui.test.js [hunk 7]
```diff
@@ -555,6 +555,210 @@ test('should only render the complete tag for task 6 item 1 when returned as com
...
+
+test('should only render task 7 item 1 as a link if recipients saved and count 1, Catalogue solution saved and count 0', async (t) => {
+  const mockOrderSummary = generateMockOrderSummary([
+    { id: 'ordering-party', status: 'complete' },
+    { id: 'supplier', status: 'complete' },
+    { id: 'commencement-date', status: 'complete' },
+    { id: 'service-recipients', status: 'complete', count: 1 },
+    { id: 'catalogue-solutions', status: 'complete', count: 0 },
```
Comment (alibombaye) [inline app/pages/task-list/existingorder/ui-tests/ui.test.js:614]: Does it not need the additional service to also be complete for it to be a link?

Comment (samMcCarthyBJSS) [reply app/pages/task-list/existingorder/ui-tests/ui.test.js:614]: not if catalogue solutions is complete with 0 added it's scenario 3

**Pull request merged**